### PR TITLE
Validation + Unit Tests for Specimen -> Tumour Grade

### PR DESCRIPTION
**Pull request by wajiha-oicr**
## Implemented the following matrix of 'codelists' as a validation function

 For the field `tumour_grade`, select a value based on tumor_grading_system according to this matrix:

Default | Gleason | Nottingham | Brain cancer | ISUP for renal cell carcinoma | Lymphoid neoplasms
-- | -- | -- | -- | -- | --
Gx - cannot be assessed | Gleason X: Gleason score cannot be determined | G1 (Low grade or well differentiated) | Grade I | Grade 1: Tumor cell nucleoli invisible or small and basophilic at 400 x magnification | Low grade or indolent NHL
G1 well differentiated/low grade | Gleason 2–6: The tumor tissue is well differentiated | G2 (Intermediate grade or moderately differentiated) | Grade II | Grade 2: Tumor cell nucleoli conspicuous at 400 x magnification but inconspicuous at 100 x magnification | High grade or aggressive NHL
G2 moderately differentiated/intermediated grade | Gleason 7: The tumor tissue is moderately differentiated | G3 (High grade or poorly differentiated) | Grade III | Grade 3: Tumor cell nucleoli eosinophilic and clearly visible at 100 x magnification |  
G3 poorly differentiated/high grade | Gleason 8–10: The tumor tissue is poorly differentiated or undifferentiated |   | Grade IV | Grade 4: Tumors showing extreme nuclear pleomorphism and/or containing tumor giant cells and/or the presence of any proportion of tumor showing sarcomatoid and/or rhabdoid dedifferentiation |  
G4 undifferentiated/high grade |   |   |   |   |  

### Additional Notes
- Both fields are required fields so any empty value would have been caught beforehand
- The values for tumour_grading_system are restricted to a codelist so it cannot deviate from those set values
- Wrote unit tests

### CODE REVIEW

File: populateReferences.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Tool to aid in testing only
+// populates the references from dictionary.json with the actual references, generates populated_dictionary.json
+// allows for easy overwrite into clinical's sample-schema.json
```
Comment (joneubank) [inline populateReferences.js]: 👌 Thank you for explaining in a comment why this file exists, I was confused.

I am cautious of giving the ability for someone to generate and upload to lectern a working schema that doesn't use references, so I'll just make sure the README is clear as to the purpose of this scritp.

File: references/validationFunctions/specimen/tumourGrade.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * 
+ * @param {Object} $row The object representing the row for a donor. Object keys represent the fields.
+ * @param {String} $field The value for the field.
+ */
+const validation = ($row,$field) => 
+    (function validate() {
+        let result = {valid: true, message: "Ok"};
+        let codeList = [];
+        switch ($row.tumour_grading_system.trim().toLowerCase()) {
+            case 'default':
```
Comment (joneubank) [inline references/validationFunctions/specimen/tumourGrade.js]: 😆 was so confused as to why `default` case was first, but this is of course the string value.

File: references/validationFunctions/specimen/tumourGrade.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                    'grade 3: tumor cell nucleoli eosinophilic and clearly visible at 100 x magnification',
+                    'grade 4: tumors showing extreme nuclear pleomorphism and/or containing tumor giant cells and/or the presence of any proportion of tumor showing sarcomatoid and/or rhabdoid dedifferentiation'
+                ];
+                break;
+            case 'lymphoid neoplasms':
+                codeList = [
+                    'low grade or indolent nhl',
+                    'high grade or aggressive nhl'
+                ];
+                break;
```
Comment (joneubank) [inline references/validationFunctions/specimen/tumourGrade.js]: I would still include a default case to handle if the codeList is expanded.

File: references/validationFunctions/specimen/tumourGrade.js (hunk 2 above)
Comment (wajiha-oicr) [reply references/validationFunctions/specimen/tumourGrade.js]: this tripped me out too lol

File: references/validationFunctions/specimen/tumourGrade.js (hunk 3 above)
Comment (wajiha-oicr) [reply references/validationFunctions/specimen/tumourGrade.js]: https://github.com/icgc-argo/argo-dictionary/pull/46/commits/83d84d2b5dcde7a379917887407398c2965d6c02

**Comment by wajiha-oicr** [comment]
@joneubank  can I get a re-review? :) :) :)

**Pull request merged**